File: kiosk/src/models/product.ts
import { IProduct } from "@/models/product.d.ts";
import User from "@/models/user.ts";

export default class Product implements IProduct {
  private _SKU!: string;
  private _description!: string;
  private _shortDescription!: string;
  private _name!: string;
  private _price!: number;
  private _stock!: number;
  private _rating!: number;
  private _reviewCount!: number;
  private _createdByAdmin!: User;
  private _startDate!: Date;
  private _endDate!: Date;
  private _createdAt!: Date;
  private _updatedAt!: Date;

  constructor() {}

  setProductFromResponseObject(obj: IProduct) {
    this._SKU = obj.SKU;
    this._name = obj.name;
    this._description = obj.description;
    this._shortDescription = obj.shortDescription;
    this._price = obj.price;
    this._stock = obj.stock;
    this._rating = obj.rating;
    this._reviewCount = obj.reviewCount;
    this._startDate = new Date(obj.startDate);
    this._endDate = new Date(obj.endDate);
    this._createdAt = new Date(obj.createdAt);
    this._updatedAt = new Date(obj.updatedAt);
  }

  get SKU(): string {
    return this._SKU;
  }

  set SKU(value: string) {
    this._SKU = value;
  }

  get description(): string {
    return this._description;
  }

  set description(value: string) {
    this._description = value;
  }

  get shortDescription(): string {
    if (this._shortDescription.length > 0)
      return this._shortDescription;

    return "No description";
  }

  set shortDescription(value: string) {
    this._shortDescription = value;
  }

  get name(): string {
    return this._name;
  }

  set name(value: string) {
    this._name = value;
  }

  get price(): number {
    return this._price;
  }

  set price(value: number) {
    this._price = value;
  }

  get stock(): number {
    return this._stock;
  }

  get rating(): number {
    return this._rating;
  }

  set rating(value: number) {
    this._rating = value;
  }

  get reviewCount(): number {
    return this._reviewCount;
  }

  set reviewCount(value: number) {
    this._reviewCount = value;
  }

  set stock(value: number) {
    this._stock = value;
  }

  get createdByAdmin(): User {
    return this._createdByAdmin;
  }

  set createdByAdmin(value: User) {
    this._createdByAdmin = value;
  }

  get startDate(): Date {
    return this._startDate;
  }

  set startDate(value: Date) {
    this._startDate = value;
  }

  get endDate(): Date {
    return this._endDate;
  }

  set endDate(value: Date) {
    this._endDate = value;
  }

  get createdAt(): Date {
    return this._createdAt;
  }

  set createdAt(value: Date) {
    this._createdAt = value;
  }

  get updatedAt(): Date {
    return this._updatedAt;
  }

  set updatedAt(value: Date) {
    this._updatedAt = value;
  }

  public getReviewDisplay(): string {
    if (this.reviewCount == 0) {
      return "No Reviews";
    }

    let display = this.reviewCount + " Review";
    if (this.reviewCount > 1)
      display += "s";
    return display;
  }
}

File: server/internal/orm/models/product/product.go
package product

import (
	"errors"
	"github.com/fedoratipper/bitkiosk/server/internal/logger"
	"github.com/fedoratipper/bitkiosk/server/internal/orm/models"
	"github.com/fedoratipper/bitkiosk/server/internal/orm/models/user"
	validation "github.com/go-ozzo/ozzo-validation/v4"
	"github.com/jinzhu/gorm"
	"regexp"
	"time"
)

type Product struct {
	models.BaseModelSoftDelete
	Name             string     `db:"name" gorm:"varchar(50)"`
	Description		 string		`db:"description" gorm:"text"`
	ShortDescription string     `db:"short_description" gorm:"varchar(255)"`
	Price            float64    `db:"price"`
	Stock            int        `db:"stock"`
	Sku              string     `db:"sku" gorm:"unique_index;index:product_sku_idx"`
	AdminId          int       `db:"admin_id"`
	StartDate        *time.Time `db:"start_date"`
	EndDate          *time.Time `db:"end_date"`
}

func (toCreate *Product) Create(db *gorm.DB) (*gorm.DB, error) {
	return models.CreateObject(toCreate, toCreate, db)
}

func (toCreate *Product) BeforeCreate(db *gorm.DB) error {
	return toCreate.Validate(db, false)
}

func (tp *Product) Validate(db *gorm.DB, toInsert bool) error {
	return validation.ValidateStruct(
		tp,
		validation.Field(&tp.Name, validation.Required, validation.Length(1, 50)),
		validation.Field(&tp.ShortDescription, validation.Required, validation.Length(1, 255)),
		validation.Field(&tp.Description, validation.Required, validation.Length(1, 4096)),
		validation.Field(&tp.AdminId, validation.By(user.ValidateAdminExists(db, "product"))),
		validation.Field(&tp.StartDate, validation.Required),
		validation.Field(&tp.EndDate, validation.By(validateEndDate(tp.StartDate))),
		validation.Field(&tp.Stock, validation.Min(0).Error("Stock count needs to be greater than 0")),
		validation.Field(&tp.Price, validation.Min( float64(0)).Error("Price needs to be greater than 0")),
		validation.Field(&tp.Sku, validation.By(validateSKUUniqueness(db, toInsert, tp.Id))),
		validation.Field(&tp.Sku, validation.By(validateSKUString())),
		)
}

func ValidateProductSkuExistence(db *gorm.DB) validation.RuleFunc {
	return func(value interface{}) error {
		sku, _ := value.(string)

		if sku == "" {
			return errors.New("Empty SKU supplied for product lookup")
		}

		product := LoadProductFromSku(sku, db)

		if product.Id == 0 {
			return errors.New("Unable to find product with SKU " + sku)
		}

		return nil
	}
}

func ValidateProductExistence(db *gorm.DB) validation.RuleFunc {
	return func(value interface{}) error {
		id, _ := value.(int)

		if id == 0 {
			return errors.New("Missing product id")
		}

		product := LoadProductWithId(id, db)

		if product.Id == 0 {
			logger.Errorfn("ValidateProductExistence", errors.New("Unable to find product with id " + string(id)))
			return errors.New("Unable to find product")
		}

		return nil
	}
}

func validateEndDate(startDate *time.Time) validation.RuleFunc {
	return func(value interface{}) error {
		endDate, _ := value.(*time.Time)

		if endDate != nil && startDate.After(*endDate) {
			return errors.New("Start date for products may not be after a given end date")
		}

		return nil
	}
}


func validateSKUString() validation.RuleFunc {
	return func(value interface{}) error {
		sku, _ := value.(string)

		var expression string = `[^\w\d\s-]`

		regex, err := regexp.Compile(expression)

		if err != nil {
			logger.Error("Unable to parse regex: \n " + expression)
			return errors.New("Unable to perform sku validation")
		}

		if len(regex.FindAllString(sku, -1)) > 0 {
			return errors.New("SKU identifiers may not contain special characters. All spaces will be replaced with hyphens")
		}

		return nil
	}
}


func validateSKUUniqueness(db *gorm.DB, toInsert bool, productId int) validation.RuleFunc {
	return func(value interface{}) error {
		sku, _ := value.(string)
		var lookupObj Product

		if !toInsert {
			db.Where("sku = ? and id != ?", sku, productId).First(&lookupObj)

			if lookupObj.Id != 0 {
				return errors.New("SKU identifiers already exists. Please enter a unique SKU for this product")
			}
		}

		return nil
	}
}